### ISSUE TITLE
Yes, please make uninstall verbs/commands, at least for fonts and simple things handled by winetricks

### ISSUE THREAD

**Issue by geekley** [Winetricks/winetricks#1863]
Firstly, I'm aware of https://wiki.winehq.org/Winetricks#How_to_remove_things_installed_by_Winetricks
> Winetricks does not provide a way to uninstall individual apps or DLLs inside a Wine prefix. This is for several reasons, but mainly because the preferred way to uninstall anything in Wine is to simply install into a fresh Wine prefix. (Yes, it would be nice to have uninstallers for everything, but I don't need it myself. Patches welcome.)
>
> If for some reason, you still don't want to fiddle at all with your Wine prefixes, Wine does offer a built-in Uninstaller program. Like the Windows "Add/Remove Programs" applet though, it only recognizes programs installed by well-behaved Windows installers that respect the registry, like InstallShield or WISE. There are no guarantees it will work with a program installed by Winetricks or other installers like .msi packages.

I understand that third-party uninstallers for complex apps/dlls may not necessarily be "well-behaved" and all you can do in this case is point to the "Add/Remove Programs" feature in Windows.

But, for simpler cases, like fonts (which I presume are handled by winetricks, not by a complex 3rd-party executable), then yes, please, there should absolutely be uninstallers.
Even if you say "you can just delete the ttf files", this isn't super-obvious - if I'm just a regular user and don't know how winetricks is implemented, I don't know **for sure** this is the ONLY thing the installer changes. The best way to give users the peace of mind that things are removed gracefully is to just make an uninstall verb/command at least for these cases - even if the command just deletes a file.
Not to mention there are packs like "corefonts" which include multiple files, which may be mixed with other ttf files added by the user - which files should be deleted in this case? What if I might have replaced them with files from my windows drive? The best way to answer is with a simple script.

**Comment by geekley** [comment]
If you do make a more "proper" uninstaller for fonts, you could ensure the deleted files were really the ones added by winetricks (not replaced by the user). For example, it could be handled like this:

- store, on installation, a simple list of font files and their MD5 hashes, and the origin package (only if installed explicitly by the user, not "dependecies"); files could have multiple entries for each manually installed package it appears in
- on uninstallation of a package, remove all entries for that package
- remove the files that had exactly 1 entry (now zero) if the hash matches; if the file is absent or has a different hash, you just assume it has been deleted manually and still remove the entry anyway

For example: `.../c:/windows/Fonts/.winetricks_installed.csv`
```csv
# Handled automatically by winetricks. DO NOT EDIT.
MD5,Filename,Package
d41d8cd98f00b204e9800998ecf8427e,corefonts.installed,corefonts
f11c0317db527bdd80fa0afa04703441,arial.ttf,arial
f11c0317db527bdd80fa0afa04703441,arial.ttf,corefonts
...
```

**Comment by austin987** [comment]
Again, it's not that simple, even in the case of fonts:
- Fonts also have registry entries (font replacements) that would need to be backed up/restored
- You'd also need to track dependencies, e.g., corefonts doesn't actually install anything, it calls the individual font verbs that then install the corresponding fonts.

Aside from that, if there's an uninstaller for fonts, users are going to request it for other verbs, i.e., dotnet* that would be way more difficult to script.

Again, however, I don't see the benefits from allowing this. It's not easy to get right, and most users don't need to uninstall stuff (i.e., if you know you don't need something, don't install it to begin with / make a backup first / reinstall the prefix). Developer time would be spent improving wine/winetricks, not trying to automate a niche case.

**Comment by geekley** [comment]
> Fonts also have registry entries (font replacements) that would need to be backed up/restored

Ah, that's unfortunate ... so I assume it's not handled by winetricks then? Is this done by a 3rd party exe/msi? Don't they provide an uninstaller too, maybe?

> You'd also need to track dependencies, e.g., corefonts doesn't actually install anything, it calls the individual font verbs that then install the corresponding fonts.

Yeah, that's why I suggested it may be done in a simple way in my second comment.

> most users don't need to uninstall stuff [...] niche case

That's an assumption, right? I highly doubt it's that uncommon.

> users are going to request it for other verbs, i.e., dotnet* that would be way more difficult to script. [...] It's not easy to get right

I agree, you're probably right on that.

> Developer time would be spent improving wine/winetricks

Yea ... that's ... the ... biggest improvement it could get ...?

> I don't see the benefits from allowing this.

Then I have to say, the "people aren't allowed to make mistakes" mentality is bad software design. "Undoability" is a really basic and important concept in software usability. Remember when you couldn't undo more than once in notepad? Didn't you hate that? Paint.NET on the other hand, gives you a full history where even things like selecting are undoable. That makes it VERY high quality software (or would if it wasn't for the fact the devs refuse to port it to linux or open the source so others can do it, but I digress). My point is: between this and adding more features/verbs or even things like a GUI, this is more important.

But I understand your points and I won't insist more or bother you if you don't want to do it. You're publishing software for free, after all.

**Comment by DaKingof** [comment]
Would it be so difficult to run the installers in a container and then have that container forward everything the installer does to wine so that it can log any changes made into a file that can be undone later? I would think it really wouldn't be too difficult to implement but I really don't know how to program either so I am just guessing here. 

I just installed .net 6 desktop LTS because a program said it needs v4.8 or later, so I tried the latest version and now I can't install older versions of .net nor uninstall v6. I don't know about you, but it seems pretty silly that now I need to nuke my whole prefix with everything in it in order to get this to work now. I have about 80 vst's in here among other things. Maybe this is a 'niche' case but it sure does feel bad that I now have to go back and re-install and setup all of my work because I can't do something as simple as uninstall a package that I can't downgrade.
